### PR TITLE
Add LLM-as-a-Judge Evaluation Framework (Phase 1) 

### PR DESCRIPTION
**Pull request by dorodb-web22**
## Add LLM-as-a-Judge Evaluation Framework (Phase 1)

**Phase 1 implementation for - #410**

## Description:

Adds native LLM-as-a-Judge evaluation metrics for RAG applications. Users can call these directly in their existing tests without external Python dependencies like Ragas or DeepEval.

## What's Included:

**New package:** `org.llm4s.eval`

| Component | Purpose |
|-----------|---------|
| `EvalContext` | Input: `(query, contexts, answer)` |
| `EvalResult` | Output: `Score(0.0-1.0) + Reasoning` |
| `Faithfulness` | Checks if answer is grounded in context |
| `AnswerRelevance` | Checks if answer addresses the query |
| `ContextPrecision` | Checks if retrieved chunks are useful |

## Usage:

```scala
val evaluator = new Evaluator(llmClient)
val result = evaluator.evaluate(
  Faithfulness(),
  EvalContext(query, contexts, answer)
)
// Right(EvalResult("Faithfulness", 0.85, true, "All claims supported"))
```

## Design Discusions:

- **Phase 1 scope**: Core metrics only, no batch runner or golden set management (as discussed in issue)
- **No new dependencies**: Uses existing LLMClient
- **Pure functions**: Stateless, no concurrency concerns
- **Follows existing patterns**: Mirrors LLMGuardrail and RAGGuardrail design

## Testing:

- 24 unit tests with mock LLM clients
- Tests cover: success cases, failure cases, error propagation, edge cases (empty context)
- All tests pass on Scala 2.13 and 3.x

CC - @vim89 @rorygraves

### CODE REVIEW

**Comment by rorygraves** [comment]
@dorodb-web22   This is a great start to the work - please address the above review comments and ping when ready.

**Comment by dorodb-web22** [comment]
@rorygraves Thanks for the detailed review! I have addressed all the issues:

**Fixed:**
- Score clamping bug - now clamps before calling `passes()`
- Fragile fallback parsing - uses `"""(\d+\.?\d*)""".r.findFirstIn()` 
- Duplicated parseResponse - extracted to `ResponseParser` (private to eval package)
- Threshold validation - added `require()` to all metric constructors
- MockLLMClient - extracted to [testutil/MockLLMClients.scala](cci:7://file:///Users/rajesh/Desktop/llm4s/modules/core/src/test/scala/org/llm4s/testutil/MockLLMClients.scala:0:0-0:0)
- Documentation - added comments for threshold choices and fail-fast behavior

**Regarding architecture refactoring:**

I would like to discuss the best approach. Two options  i see:

1. **Shared base trait** - Create `LLMJudge` trait in `agent.guardrails.llmjudge` that both guardrails and eval metrics extend. ResponseParser would move there.

2. **Composition** - Keep eval metrics separate but have them delegate to guardrail implementations internally (e.g., `Faithfulness` wraps `GroundingGuardrail`).

Option 1 seems cleaner but requires more refactoring. Option 2 is quicker but adds indirection. Happy to implement whichever approach you prefer, or discuss further.

File: modules/core/src/main/scala/org/llm4s/eval/metrics/Faithfulness.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.llm4s.llmconnect.model._
+import org.llm4s.types.Result
+
+/**
+ * Faithfulness metric - measures if the answer is grounded in the context.
+ *
+ * @param threshold Score threshold for passing (default: 0.7). Must be in [0.0, 1.0].
+ */
+class Faithfulness(override val threshold: Double = 0.7) extends EvalMetric {
+  require(threshold >= 0.0 && threshold <= 1.0, s"Threshold must be in [0.0, 1.0], got $threshold")
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/metrics/Faithfulness.scala]: This uses `require(...)` in a public core class. That throws an exception and breaks the **errors are values** rule.

Fix: use a safe constructor (companion `def fromThreshold(th: Double): Result[Faithfulness]`) or validate inside `evaluate` and return `Left(ValidationError/ConfigurationError)`.

File: modules/core/src/main/scala/org/llm4s/eval/ResponseParser.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    lines
+      .find(_.startsWith(s"$fieldName:"))
+      .map(_.stripPrefix(s"$fieldName:").trim)
+      .getOrElse("")
+  }
+
+  /**
+   * Fallback: extract first valid decimal number from response.
+   */
+  def extractFirstNumber(response: String): Option[Double] =
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/ResponseParser.scala:37]: Bug: `extractFirstNumber` does not parse signs. `"-0.2"` becomes `0.2` and can incorrectly pass validation.

Fix: parse an optional sign (e.g. `(-?\d+\.?\d*)`) and then validate `[0,1]`.

File: modules/core/src/main/scala/org/llm4s/eval/Evaluator.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+class Evaluator(val llmClient: LLMClient) {
+
+  def evaluate(metric: EvalMetric, context: EvalContext): Result[EvalResult] =
+    metric.evaluate(context, llmClient)
+
+  /**
+   * Evaluate all metrics against a single context.
+   * Stops on first error (fail-fast).
+   */
+  def evaluateAll(metrics: Seq[EvalMetric], context: EvalContext): Result[Seq[EvalResult]] =
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/Evaluator.scala:18]: This appends with `results :+ result` inside a fold. That is O(n^2) and easy to avoid.

Fix: accumulate in a `List` (prepend) and reverse at the end, or use a builder.

File: modules/core/src/test/scala/org/llm4s/eval/EvaluatorSpec.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import org.llm4s.error.NetworkError
+import org.llm4s.llmconnect.LLMClient
+import org.llm4s.llmconnect.model._
+import org.llm4s.types.Result
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class EvaluatorSpec extends AnyFlatSpec with Matchers {
+
+  class MockLLMClient(response: String) extends LLMClient {
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/eval/EvaluatorSpec.scala]: You already added `org.llm4s.testutil.MockLLMClient`, but this spec defines another local MockLLMClient.

Fix: reuse the shared test util to avoid duplicate behavior and reduce future edits.

File: modules/core/src/main/scala/org/llm4s/eval/EvalMetric.scala [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package org.llm4s.eval
+
+import org.llm4s.llmconnect.LLMClient
+import org.llm4s.types.Result
+
+/**
+ * Base trait for evaluation metrics.
+ */
+trait EvalMetric {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/EvalMetric.scala:9]: This metric duplicates existing guardrail behavior (prompt → LLM → parse score → threshold).

Fix: extract shared judge logic and have both guardrails + eval metrics call into it. Otherwise we will drift and fix bugs twice.

**Comment by dorodb-web22** [comment]
@vim89 Thanks for the detailed review. I have addressed all your technical feedback:

**Fixed (Code & Correctness):**
-  **Safe Constructors**: Removed `require()` from core classes. `Faithfulness(0.7)` now returns `Result[Faithfulness]`. Added `unsafe()` only for tests.
-  **Performance**: Optimised `evaluateAll` to O(n) using prepend + reverse builder pattern.
-  **Parsing**: `ResponseParser` now correctly handles negative numbers (e.g. `-0.2`).
-  **Testing**: Removed duplicate `MockLLMClient` in generic specs; now importing from `testutil`.

**Logic & Architecture:**
- I agree that sharing judge logic between guardrails and eval metrics is the right long-term move.
- However, to keep this PR focused on fixing the critical bugs in the current implementation, I propose landing these fixes first.
- i am  happy to open a follow-up PR dedicated to `modules/eval` refactoring and extracting a shared `LLMJudge` trait, as that involves touching existing guardrail code which might be risky to bundle here.

Ready for another look @vim89 @rorygraves

File: modules/core/src/main/scala/org/llm4s/eval/EvalMetric.scala (hunk 5 above)
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/EvalMetric.scala:9]: Still not addressed: this creates a second judge flow in core, while similar logic already exists in rag evaluation and guardrails. Please extract one shared judge engine and have both paths call it.
Guideline principle: Prefer one shared core path over parallel implementations.

File: modules/core/src/main/scala/org/llm4s/eval/metrics/Faithfulness.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+object Faithfulness {
+  def apply(threshold: Double = 0.7): Result[Faithfulness] =
+    if (threshold >= 0.0 && threshold <= 1.0) Right(new Faithfulness(threshold))
+    else Left(ConfigurationError(s"Threshold must be in [0.0, 1.0], got $threshold"))
+
+  def strict: Result[Faithfulness]  = apply(0.9)
+  def lenient: Result[Faithfulness] = apply(0.5)
+
+  /** Unsafe constructor for tests/convenience - throws on invalid threshold */
+  def unsafe(threshold: Double = 0.7): Faithfulness =
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/metrics/Faithfulness.scala]: Still not addressed: this public unsafe constructor throws exceptions in core code. Please keep errors as Result values and move throwing helpers to test-only code.
Guideline principle: Core domain code should return typed errors, not throw.

File: modules/core/src/main/scala/org/llm4s/eval/metrics/AnswerRelevance.scala [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      case Some(rawScore) =>
+        val clampedScore = Math.max(0.0, Math.min(1.0, rawScore))
+        Right(EvalResult(name, clampedScore, passes(clampedScore), explanation))
+
+      case None =>
+        ResponseParser.extractFirstNumber(response) match {
+          case Some(rawScore) if rawScore >= 0.0 && rawScore <= 1.0 =>
+            Right(EvalResult(name, rawScore, passes(rawScore), ""))
+          case _ =>
+            Left(ConfigurationError(s"Could not parse answer relevance score: ${response.take(100)}"))
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/eval/metrics/AnswerRelevance.scala]: Parse failure is not a config problem. Please return ValidationError or ProcessingError here so callers can handle it correctly.
Guideline principle: Error types should match the real failure source.

**Comment by vim89** [comment]
This is architecturally concerning in its current shape.
It adds another judge path in core while similar logic already exists in guardrails and RAG evaluation.
That creates long-term drift and duplicate bug-fix cost.
Please stop adding surface area and extract one shared judge engine first.
Without that, this is not a clean value-add.

@AnshumanAI please review this.

**Comment by dorodb-web22** [comment]
@vim89 Thanks for the follow-up. i have addressed the remaining points:

1. **Error Types**: Core metrics now return `ValidationError` for parsing failures (instead of `ConfigurationError`), matching the actual failure domain.
2. **Safe Core**: Completely removed `unsafe()` constructors from core logic.
3. **Test Infrastructure**: Created `org.llm4s.testutil.UnsafeMetrics` for test-only convenience helpers.

Regarding architecture: I agree on the shared judge engine. Though ,i prefer to take up that refactor as a dedicated follow-up after @AnshumanAI's review, to avoid destabilising existing guardrails in this bug-fix PR

**Comment by rorygraves** [comment]
Hi @dorodb-web22 — thanks for continuing to iterate on this. You've done good work addressing the technical issues (safe constructors, O(n) evaluateAll, negative parsing, ValidationError). However, after a deeper review I need to flag a fundamental problem that wasn't fully visible before.

## The core issue: these metrics already exist

The codebase already has a full RAG evaluation framework at `org.llm4s.rag.evaluation.metrics` with **all three metrics** this PR adds:

| This PR (`org.llm4s.eval`) | Already exists (`org.llm4s.rag.evaluation.metrics`) |
|---|---|
| `Faithfulness` — single LLM call, SCORE: line parsing | `Faithfulness` — multi-step claim extraction + batch verification with detailed breakdowns |
| `AnswerRelevance` — single LLM call, SCORE: line parsing | `AnswerRelevancy` — embedding-based cosine similarity across generated questions |
| `ContextPrecision` — single LLM call, SCORE: line parsing | `ContextPrecision` — Average Precision ranking metric with per-context relevance assessment |

On top of that, the guardrails package has overlapping judge logic (`LLMFactualityGuardrail`, `GroundingGuardrail`, `ContextRelevanceGuardrail`).

This means the PR creates a **third parallel system** for the same evaluations, using simpler (less accurate) approaches than what already exists. This is exactly the architectural concern vim89 raised across both review rounds.

## Additional issue: unrelated build file changes

The PR includes whitespace-only reformatting of `build.sbt`, `plugins.sbt`, and `WorkspaceRunnerDocker.scala` — these are unrelated to the eval feature and shouldn't be bundled here.

## Recommendation

We're going to close this PR since the functionality already exists in the codebase. If you'd like to continue contributing in this area, here are some high-value alternatives:

1. **Add tests for the existing `rag.evaluation.metrics`** — they could use better coverage
2. **Extract the shared judge engine** vim89 suggested — common parsing/prompting logic used by both guardrails and RAG evaluation
3. **Improve the existing metrics** — e.g., better error handling, performance optimizations, or additional metrics not yet covered

Thanks for the effort on this — the code quality improved significantly across the iterations, and those skills will be valuable on a follow-up that builds on the existing infrastructure rather than duplicating it.

**Comment by AnshumanAI** [comment]
Agree with @vim89 and @rorygraves 's comments, however i propose we can scavenge tests , if you'd like.

Please create a new PR dedicated to the test additions so they can be reviewed separately.

**Comment by rorygraves** [comment]
@dorodb-web22  - as per above - this work duplictes existing parts but it would be great if you extract the test addtions into a new PR.  Closing for now.

**Comment by dorodb-web22** [comment]
@rorygraves @vim89 @AnshumanAI - Thank you for the thorough reviews and patience through this process.

I see now that `org.llm4s.rag.evaluation.metrics` already covers this functionality. I appreciate you pointing that out, as well as the architectural feedback regarding the shared judge engine.

I appreciate the detailed feedback on code quality, error handling, and architecture principles throughout this process. Those insights are valuable and I learned a lot from the review rounds.

*****Next** **Steps:*****
I'll follow your suggestions,

1. Extract the test utilities (`MockLLMClients`, `UnsafeMetrics`) as a separate PR to add test coverage for the existing evaluation metrics
2. After that, explore the shared judge engine refactoring you both mentioned

Thanks again for the constructive guidance. Looking forward to contributing in a way that builds on the existing architecture.

**Comment by dorodb-web22** [comment]
Raised the follow-up PR here: #697

Testing coverage for RAG evaluation metrics + shared MockLLMClients utility.
